Have gatherFragmentImportsForContext include resolved filenames of imported files

grconrad opened the pull request:
The existing `gatherFragmentImportsForContext` is missing something I need: the resolved path of each imported file.

I'm writing a lint rule that needs to know _where_ imported fragments were imported from, i.e. the resolved path of each imported file alongside the existing map of fragment names to definitions contained in that imported file.

I believe we have two options here:

1. Augment the existing information to add the resolvedPath per import. I think this would have to a breaking change, as the shape of the entries in the map keyed by line number would need to change to add the resolvedPath somewhere.
2. Add a parallel implementation that returns data in a slightly different shape.

This PR uses option 2 and adds new functions to support that, i.e. `gatherAttributedFragmentImportsForContext` and others. I realize this is a lot of duplication. Happy to hear ideas, including option 1 (change the existing `gatherFragmentImportsForContext` and doing a major bump) or others.

Commit:
Expose a way to get summarized imports with resolvedPath

stefanpenner commented:
Thanks @grconrad, could you elaborate on what would be breaking about (1) ?

stefanpenner commented on lib/gather-attributed-fragment-imports.js:
```diff
@@ -0,0 +1,60 @@
+    for (const fragment of fragmentParserGenerator(line)) {
+      fragmentDefinitionsByName.set(fragment.name.value, fragment);
+    }
+    const fileFragmentsInfo = {
```
Is this the breaking change, if we took (1)?

grconrad replied:
resolving since we discussed

stefanpenner commented on lib/inline-imports.js:
```diff
@@ -59,7 +59,7 @@ function* linesWithInlinedImportsOf(
         visited,
       );
-      yield { line, match: true, lineNumber };
+      yield { line, match: true, lineNumber, resolvedPath };
```
`resolvedPath` may be an OK name, but I am curious if there is a better name that is more specific

stefanpenner commented:
The least invasive, and probably pretty reasonable approach is to tack `filename` onto each `fragment.loc.filename`. Although it is a little odd to "pollute" a structure that isn't ours. If they choose to add `filename` we wouldn't conflict as we are literally writing the valid filename to that position.

Again although odd, I do feel it is perfectly reasonable in this scenario.

grconrad commented:
Thanks @stefanpenner . To capture our offline discussion here, the breaking change would be adding new fields inside the values in the Map to include `resolvedPath`. I agree that all the duplication to avoid a major bump is bad. We could major bump and change the structure of the data, but first I'll explore adding filename inside `loc` property of the FragmentDefinition node, i.e. to populate `fooFragment.loc.filename`. Curiously, `loc` has `start` and `end` (which both have column and line numbers) but not the file path.

Commit:
Switch to simpler approach to eliminate duplication
Write resolved file path to .loc.filename on fragment definition node
Adjust fake parsers (used in tests) to set loc

grconrad commented:
@stefanpenner Updated per the simpler approach. Let me know what you think. I did have to update some tests.

stefanpenner commented on lib/gather-fragment-imports.js:
```diff
@@ -1,6 +1,6 @@
 const path = require('path');
 
-const inlineImports = require('@graphql-fragment-import/lib/inline-imports');
```
do we need this change?

grconrad replied:
So, when I ran `yarn test` from the top level of the project, it actually ran the linter and that linter complained about this line. :) Not a big deal but I imagine a project referring to itself this way is pretty uncommon.

stefanpenner replied:
Interesting, Feel free to make the change then.

grconrad replied:
With the existing code, the lint error that fails `yarn test` or `yarn lint` is:
```
"@graphql-fragment-import/lib" is extraneous.
```
eslint([node/no-extraneous-require](https://github.com/mysticatea/eslint-plugin-node/blob/v11.1.0/docs/rules/no-extraneous-require.md))

So I figured I'd clean this up by using a relative path

stefanpenner commented on lib/inline-imports.js:
```diff
@@ -89,15 +89,15 @@ function inlineImports(fileContents, options = {}, visited = new Set()) {
     // non-matching lines
     if (match) {
-      result.set(lineNumber, line);
+      result.set(lineNumber, { resolvedPath, line });
```
Could we use `filename` ?

grconrad replied:
Sure, I can do that. It will actually be the full file path (not just base name) but that's the same as in `fragmentDefinition.loc.filename` so I can carry the `filename` terminology throughout.

grconrad replied:
Done

stefanpenner commented on lib/tests/gather-fragment-imports-test.js:
```diff
@@ -102,7 +103,20 @@ fragment FromDependency on User {
+          new Map([
+            [
+              'apple',
+              { name: { value: 'apple' }, loc: { filename: path.join(basedir, 'apple.graphql') } },
```
LGTM

stefanpenner commented:
I like this approach, how do you feel about it?

Commit:
Review feedback: resolvedPath --> filename
